### PR TITLE
Rover: add marvelmind beacon support

### PR DESCRIPTION
**Pull request by amilcarlucas**
This almost adds MarvelMind (indoors Ultrasound-based GPS) beacon support to Rover.
But it does uses the library, so other vehicles should be possible as well

### CODE REVIEW

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+    uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Beacon,
+            0);
+    if (uart != nullptr) {
+        uart->begin(
+                serial_manager.find_baudrate(
+                        AP_SerialManager::SerialProtocol_Beacon, 0));
+
+    //variables for the receive thread
+        hedge = new MarvelmindHedge();
+        if(hedge)
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: should be if (hedge) {

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+        {
+            createMarvelmindHedge();
+            Parse_State = RECV_HDR; // current state of receive data
+            nBytesInBlockReceived = 0; // bytes received
+            dataId = 0;
+            startMarvelmindHedge();
+        }else{
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: } else {

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    }
+
+}
+
+void AP_Beacon_Marvelmind::getOrAllocBeacon(struct StationaryBeaconPosition &b,
+        uint8_t address) {
+    uint8_t i;
+    uint8_t n_used = hedge->positionsBeacons.numBeacons;
+
+    if (n_used != 0)
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: miss {

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+        for (i = 0; i < n_used; i++) {
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: uint8_t i should be declared here ! not before !

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+            //Initialising beacon failed
+        }
+    }
+
+}
+
+void AP_Beacon_Marvelmind::getOrAllocBeacon(struct StationaryBeaconPosition &b,
+        uint8_t address) {
+    uint8_t i;
+    uint8_t n_used = hedge->positionsBeacons.numBeacons;
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+
+    if (n_used != 0)
+        for (i = 0; i < n_used; i++) {
+            if (hedge->positionsBeacons.beacons[i].address == address) {
+                b = hedge->positionsBeacons.beacons[i];
+                return;
+            }
+        }
+
+    if (n_used >= (MAX_STATIONARY_BEACONS - 1))
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: miss {}

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+}
+
+void AP_Beacon_Marvelmind::process_beacons_positions_datagram(
+        struct StationaryBeaconPosition &b) {
+    uint8_t n = input_buffer[5]; // number of beacons in packet
+    uint8_t i, ofs;
+    uint8_t address;
+    int16_t x, y, z;
+
+    if ((1 + n * 8) != input_buffer[4])
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: miss {}

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+        return; // incorrect size
+
+    for (i = 0; i < n; i++) {
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare uint8_t i here not before

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+        ofs = 6 + i * 8;
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare ofs here not before. and make it const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+
+        address = input_buffer[ofs + 0];
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare address here not before. and make it const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+        x = input_buffer[ofs + 1] | (((uint16_t) input_buffer[ofs + 2]) << 8);
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare x here not before. and make it const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+        y = input_buffer[ofs + 3] | (((uint16_t) input_buffer[ofs + 4]) << 8);
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare y here not before. and make it const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp (hunk 10 above)
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: is "+ 0" necessary ?

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+        z = input_buffer[ofs + 5] | (((uint16_t) input_buffer[ofs + 6]) << 8);
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare z here not before. and make it const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+
+        getOrAllocBeacon(b, address);
+        if (b.address != NULL) {
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: nullptr

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+}
+
+void AP_Beacon_Marvelmind::process_beacons_positions_highres_datagram(
+        struct StationaryBeaconPosition &b) {
+    uint8_t n = input_buffer[5]; // number of beacons in packet
+    uint8_t i, ofs;
+    uint8_t address;
+    int32_t x, y, z;
+
+    if ((1 + n * 14) != input_buffer[4])
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: miss {}

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+        return; // incorrect size
+
+    for (i = 0; i < n; i++) {
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare uint8_t i here not before

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 17]
```diff
@@ -0,0 +1,399 @@
...
+            b.highResolution = false;
+
+            hedge->positionsBeacons.updated = true;
+        }
+    }
+}
+
+void AP_Beacon_Marvelmind::process_beacons_positions_highres_datagram(
+        struct StationaryBeaconPosition &b) {
+    uint8_t n = input_buffer[5]; // number of beacons in packet
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 18]
```diff
@@ -0,0 +1,399 @@
...
+    if (n_used >= (MAX_STATIONARY_BEACONS - 1))
+        return;
+
+    hedge->positionsBeacons.numBeacons = (n_used + 1);
+    b = hedge->positionsBeacons.beacons[n_used];
+}
+
+void AP_Beacon_Marvelmind::process_beacons_positions_datagram(
+        struct StationaryBeaconPosition &b) {
+    uint8_t n = input_buffer[5]; // number of beacons in packet
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 19]
```diff
@@ -0,0 +1,399 @@
...
+    uint8_t i, ofs;
+    uint8_t address;
+    int32_t x, y, z;
+
+    if ((1 + n * 14) != input_buffer[4])
+        return; // incorrect size
+
+    for (i = 0; i < n; i++) {
+        ofs = 6 + i * 14;
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare ofs here not before and const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 20]
```diff
@@ -0,0 +1,399 @@
...
+
+        address = input_buffer[ofs + 0];
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare ofs here not before and const.
"+ 0" necessary ?

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 21]
```diff
@@ -0,0 +1,399 @@
...
+        x = input_buffer[ofs + 1] | (((uint32_t) input_buffer[ofs + 2]) << 8)
+                | (((uint32_t) input_buffer[ofs + 3]) << 16)
+                | (((uint32_t) input_buffer[ofs + 4]) << 24);
+        y = input_buffer[ofs + 5] | (((uint32_t) input_buffer[ofs + 6]) << 8)
+                | (((uint32_t) input_buffer[ofs + 7]) << 16)
+                | (((uint32_t) input_buffer[ofs + 8]) << 24);
+        z = input_buffer[ofs + 9] | (((uint32_t) input_buffer[ofs + 10]) << 8)
+                | (((uint32_t) input_buffer[ofs + 11]) << 16)
+                | (((uint32_t) input_buffer[ofs + 12]) << 24);
+
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare x, y, z here not before and const.

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 22]
```diff
@@ -0,0 +1,399 @@
...
+        getOrAllocBeacon(b, address);
+        if (b.address != NULL) {
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: nullptr

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 23]
```diff
@@ -0,0 +1,399 @@
...
+}
+
+uint8_t AP_Beacon_Marvelmind::markPositionReady() {
+    uint8_t ind = hedge->lastValues_next;
+    uint8_t indCur = ind;
+
+    hedge->positionBuffer[ind].ready = true;
+    hedge->positionBuffer[ind].processed = false;
+    ind++;
+    if (ind >= hedge->maxBufferedPositions)
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: miss {}

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 24]
```diff
@@ -0,0 +1,399 @@
...
+        ind = 0;
+    if (hedge->lastValuesCount_ < hedge->maxBufferedPositions)
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: miss {}

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 25]
```diff
@@ -0,0 +1,399 @@
...
+        hedge->lastValuesCount_++;
+    hedge->haveNewValues_ = true;
+
+    hedge->lastValues_next = ind;
+
+    return indCur;
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: indCur is return without have been used ? Is something missing ?
if not, make it const !

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 26]
```diff
@@ -0,0 +1,399 @@
...
+void AP_Beacon_Marvelmind::process_position_datagram(struct PositionValue &p) {
+    uint8_t ind = hedge->lastValues_next;
+
+    hedge->positionBuffer[ind].address = input_buffer[16];
+    hedge->positionBuffer[ind].timestamp = input_buffer[5]
+            | (((uint32_t) input_buffer[6]) << 8)
+            | (((uint32_t) input_buffer[7]) << 16)
+            | (((uint32_t) input_buffer[8]) << 24);
+
+    int16_t vx = input_buffer[9] | (((uint16_t) input_buffer[10]) << 8);
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 27]
```diff
@@ -0,0 +1,399 @@
...
+    hedge->positionBuffer[ind].x = vx * 10; // millimeters
+
+    int16_t vy = input_buffer[11] | (((uint16_t) input_buffer[12]) << 8);
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 28]
```diff
@@ -0,0 +1,399 @@
...
+    hedge->positionBuffer[ind].y = vy * 10; // millimeters
+
+    int16_t vz = input_buffer[13] | (((uint16_t) input_buffer[14]) << 8);
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 29]
```diff
@@ -0,0 +1,399 @@
...
+        struct PositionValue &p) {
+    uint8_t ind = hedge->lastValues_next;
+
+    hedge->positionBuffer[ind].address = input_buffer[22];
+    hedge->positionBuffer[ind].timestamp = input_buffer[5]
+            | (((uint32_t) input_buffer[6]) << 8)
+            | (((uint32_t) input_buffer[7]) << 16)
+            | (((uint32_t) input_buffer[8]) << 24);
+
+    int32_t vx = input_buffer[9] | (((uint32_t) input_buffer[10]) << 8)
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 30]
```diff
@@ -0,0 +1,399 @@
...
+            | (((uint32_t) input_buffer[11]) << 16)
+            | (((uint32_t) input_buffer[12]) << 24);
+    hedge->positionBuffer[ind].x = vx;
+
+    int32_t vy = input_buffer[13] | (((uint32_t) input_buffer[14]) << 8)
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 31]
```diff
@@ -0,0 +1,399 @@
...
+            | (((uint32_t) input_buffer[15]) << 16)
+            | (((uint32_t) input_buffer[16]) << 24);
+    hedge->positionBuffer[ind].y = vy;
+
+    int32_t vz = input_buffer[17] | (((uint32_t) input_buffer[18]) << 8)
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: const

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 32]
```diff
@@ -0,0 +1,399 @@
...
+
+    ind = markPositionReady();
+
+    p = hedge->positionBuffer[ind];
+}
+
+uint16_t AP_Beacon_Marvelmind::CalcCrcModbus_(uint8_t * buf, int len) {
+    uint16_t crc = 0xFFFF;
+    int pos;
+    for (pos = 0; pos < len; pos++) {
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare pos here not before !

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 33]
```diff
@@ -0,0 +1,399 @@
...
+        crc ^= (uint16_t) buf[pos]; // XOR byte into least sig. byte of crc
+        int i;
+        for (i = 8; i != 0; i--) // Loop over each bit
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare int here not before !

Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: for (uint8_t i = 8; i != 0; i--) {

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 34]
```diff
@@ -0,0 +1,399 @@
...
+                {
+            if ((crc & 0x0001) != 0) // If the LSB is set
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: if ((crc & 0x0001) != 0) {

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 35]
```diff
@@ -0,0 +1,399 @@
...
+                    {
+                crc >>= 1; // Shift right and XOR 0xA001
+                crc ^= 0xA001;
+            } else
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: } else {

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 36]
```diff
@@ -0,0 +1,399 @@
...
+    }
+    return crc;
+}
+
+void AP_Beacon_Marvelmind::createMarvelmindHedge() {
+
+    hedge->maxBufferedPositions = 3;
+    hedge->positionBuffer = NULL;
+    hedge->verbose = false;
+    hedge->receiveDataCallback = NULL;
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: nullptr

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 37]
```diff
@@ -0,0 +1,399 @@
...
+    uint8_t i;
+    hedge->positionBuffer = (PositionValue*) malloc(
+            sizeof(struct PositionValue) * hedge->maxBufferedPositions);
+    if (hedge->positionBuffer == NULL) {
+        if (hedge->verbose)
+            puts("Not enough memory");
+        hedge->terminationRequired = true;
+        return;
+    }
+    for (i = 0; i < hedge->maxBufferedPositions; i++) {
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: declare uint8_t i here not before

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 38]
```diff
@@ -0,0 +1,399 @@
...
+                crc >>= 1; // Just shift right
+        }
+    }
+    return crc;
+}
+
+void AP_Beacon_Marvelmind::createMarvelmindHedge() {
+
+    hedge->maxBufferedPositions = 3;
+    hedge->positionBuffer = NULL;
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: nullptr

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 39]
```diff
@@ -0,0 +1,399 @@
...
+    if (uart == nullptr) {
+        return;
+    }
+
+    // read any available characters
+    int32_t nBytesRead = uart->available();
+
+    uint8_t receivedChar = 0;
+    bool readSuccessed = true;
+    if (nBytesRead < 0)
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: miss {}

Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: maybe simply with  if (nBytesRead < 0) { 
return; 
}
and get ride of readSuccessed, no ?

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 40]
```diff
@@ -0,0 +1,399 @@
...
+                    case BEACONS_POSITIONS_DATAGRAM_ID:
+                    case BEACONS_POSITIONS_DATAGRAM_HIGHRES_ID:
+                        goodByte = true;
+                        break;
+                    case POSITION_DATAGRAM_HIGHRES_ID: {
+                        goodByte = (receivedChar == 0x16);
+                    }
+                        break;
+                    }
+                    if (goodByte)
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: miss {}

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 41]
```diff
@@ -0,0 +1,399 @@
...
+                                    || (dataId == POSITION_DATAGRAM_HIGHRES_ID)
+                                    || (dataId
+                                            == BEACONS_POSITIONS_DATAGRAM_HIGHRES_ID);
+                    break;
+                case 4: {
+                    switch (dataId) {
+                    case POSITION_DATAGRAM_ID: {
+                        goodByte = (receivedChar == 0x10);
+                    }
+                        break;
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: i think the break should be in the enclosing {}

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 42]
```diff
@@ -0,0 +1,399 @@
...
+                    case BEACONS_POSITIONS_DATAGRAM_ID:
+                    case BEACONS_POSITIONS_DATAGRAM_HIGHRES_ID:
+                        goodByte = true;
+                        break;
+                    case POSITION_DATAGRAM_HIGHRES_ID: {
+                        goodByte = (receivedChar == 0x16);
+                    }
+                        break;
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: i think the break should be in the enclosing {}

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp [hunk 43]
```diff
@@ -0,0 +1,399 @@
...
+                    }
+                    if (goodByte)
+                        Parse_State = RECV_DGRAM;
+                }
+                    break;
```
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: i think the break should be in the enclosing {}

**Comment by khancyr** [comment]
@amilcarlucas I add a few comments about style ^^. More generally, we are using C++ so it is better to reduce the variable scope. You don't need to predeclare the variable on function start, it is better to do it when needed. 
Remember to alway use {} on one line if ... It will prevent stupid mistake !

**Comment by karthikdesai1** [comment]
@khancyr will do this soon! I just adapted Marvelmind C API to work inside Ardupilot.
Problem is that we wanted to enable Beacon class for the Rover and listen to the position from the beacon on the serial port.
With this code, i was able to set the BCN parameters for the rover, but this wasn't reflected for the beacon object! I think there is a parameter or header file missing some where. I already contacted @rmackay9 regarding this.

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp (hunk 14 above)
Comment (karthikdesai1) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: /ardupilot/libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp:81:26: error: invalid operands of types 'uint8_t {aka unsigned char}' and 'std::nullptr_t' to binary 'operator!='

i receive this error

Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: That why NULL is tricky ... address is uint8_t, so it just need to be b.address != 0 ^^

Comment (karthikdesai1) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: *b->address != nullptr makes sense?

File: libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp (hunk 22 above)
Comment (khancyr) [inline libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp]: b.address != 0

**Comment by rmackay9** [comment]
I hope/plan to review this tomorrow. sorry for the delay.

**Comment by rmackay9** [comment]
@amilcarlucas,
Thanks for adding this feature!

I've added a couple of commits on top of yours in this repo: https://github.com/rmackay9/rmackay9-ardupilot/commits/amilcar-marvelmind

The biggest changes are:

- rover updates at only 50hz, increased priority of beacon.update
- changed camel case to underscore
- moved Beacon parameters to be after AFS (we can't change the eeprom position of existing parameters without corrupting people's parameters).

Two issues remaining:

- I think there's the possibility of the input_buffer overflowing.  The buffer is 255 characters (I reduced it by 1) but there's nothing directly stopping it from overwriting the end of the buffer as far as I can tell.
- instead of duplicating the beacon dataflash logging in rover, we should move Copter's beacon logging to the DataFlash class.  That may mean we need to add "#include AP_Beacon/AP_Beacon" in a ton of places in the vehicle code, example sketches and replay but better to do that than to duplicate the code I think.

Overall looking pretty good!

**Comment by karthikdesai1** [comment]
Hello @rmackay9,
Thank you!! Will apply your changes and push with with @amilcarlucas by the end of the day! 😄

**Comment by amilcarlucas** [comment]
Cleaned-up, rebased and squashed

**Comment by gmorph** [comment]
I don't know why its complaining about Parameters.h - the change looks fine to me.

File: libraries/AP_Beacon/AP_Beacon.cpp [hunk 44]
```diff
@@ -86,6 +87,8 @@ void AP_Beacon::init(void)
     // create backend
     if (_type == AP_BeaconType_Pozyx) {
         _driver = new AP_Beacon_Pozyx(*this, serial_manager);
+    } else if(_type == AP_BeaconType_Marvelmind){
```
Comment (gmorph) [inline libraries/AP_Beacon/AP_Beacon.cpp]: Space after the if would be good - so "if ("

**Comment by rmackay9** [comment]
OK, merged!

**Comment by rmackay9** [comment]
Ah, forgot to mention, I made the small fixes that @gmorph requested as part of the merge.

**Comment by rmackay9** [comment]
@amilcarlucas, it would be great to get some instructions on the wiki as to how to set this up.  I'm happy to help you with that.

**Pull request closed without merging**